# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [29]:
#load python packages
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
from sklearn.cluster import KMeans

os.getcwd()


'C:\\Users\\nmur1\\Google Drive\\Springboard\\Guided Capstone\\data'

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [30]:
path = r'C:\Users\nmur1\Google Drive\Springboard\Guided Capstone\data'
path

'C:\\Users\\nmur1\\Google Drive\\Springboard\\Guided Capstone\\data'

In [31]:
os.chdir(path)
df = pd.read_csv('step2_output.csv')
df.head()
df.drop(columns = "Unnamed: 0", inplace = True)


In [32]:
from sklearn.cluster import KMeans
x = df.drop(['Name','state', 'Region'], axis =1)

kmeans3 = KMeans(n_clusters=3)
y_kmeans3 = kmeans3.fit_predict(x)

df['cluster'] = y_kmeans3
df.head()
df.to_clipboard()

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [33]:
df_state = df[['state']]

df2 = pd.concat([df.drop(df_state, axis = 1), pd.get_dummies(df_state)], axis =1)
df2.drop(['Region'], axis = 1, inplace = True)
df2.head()

Name  summit_elev  vertical_drop  base_elev  trams  \
0       Alyeska Resort         3939           2500        250      1   
1  Eaglecrest Ski Area         2600           1540       1200      0   
2     Hilltop Ski Area         2090            294       1796      0   
3     Arizona Snowbowl        11500           2300       9200      0   
4  Sunrise Park Resort        11100           1800       9200      0   

   fastEight  fastSixes  fastQuads  quad  triple  ...  state_Rhode Island  \
0        0.0          0          2     2       0  ...                   0   
1        0.0          0          0     0       0  ...                   0   
2        0.0          0          0     0       1  ...                   0   
3        0.0          1          0     2       2  ...                   0   
4        0.0          0          1     2       3  ...                   0   

   state_South Dakota  state_Tennessee  state_Utah  state_Vermont  \
0                   0                0           0              0   
1                   0                0           0              0   
2                   0                0           0              0   
3                   0                0           0              0   
4                   0                0           0              0   

   state_Virginia  state_Washington  state_West Virginia  state_Wisconsin  \
0               0                 0                    0                0   
1               0                 0                    0                0   
2               0                 0                    0                0   
3               0                 0                    0                0   
4               0                 0                    0                0   

   state_Wyoming  
0              0  
1              0  
2              0  
3              0  
4              0  

[5 rows x 61 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [34]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X= df2.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y= df2.AdultWeekend 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

pd.DataFrame(X_scaled).to_clipboard()


## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [35]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

#pd.DataFrame(y_train).to_clipboard()
#pd.DataFrame(y_train).to_clipboard()

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [36]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error, mean_squared_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

lm.intercept_

64.12473706791931

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [37]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)


## Review Model Outcomes — Iterate over additional models as needed

In [38]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model

In [39]:


from math import sqrt

rmse = sqrt(mean_squared_error(y_test, y_pred))
mae = mean_absolute_error(y_test, y_pred)
evs = explained_variance_score(y_test, y_pred)

print(mae)
print(evs)

5.163452216083286
0.9307211181167558


**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [40]:
print(model.intercept_)

64.12473706791931


**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [41]:
coef = np.round(lm.coef_,2)

In [42]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(coef), X.columns, columns=['Coefficient']).sort_values(by='Coefficient', ascending = False).head(10)


Coefficient
total_chairs    8.599293e+12
fastQuads       3.260011e+12
surface         3.054386e+12
double          2.691638e+12
triple          2.401126e+12
quad            1.946024e+12
fastSixes       9.664321e+11
state_New York  8.957149e+11
state_Michigan  8.453117e+11
trams           8.303853e+11

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [43]:
#Scale the model

#drop state, name, adulstweekend
X2 = df.drop(['Name','AdultWeekend', 'state', 'Region'], axis=1)
#set y
y= df.AdultWeekend 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X2)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X2) 

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train2, X_test2, y_train2, y_test2 = train_test_split(X_scaled,y, test_size=0.25, random_state=1)

pd.DataFrame(X_test2, columns = X2.columns).to_clipboard()

In [44]:
#rerun regression
lm2 = linear_model.LinearRegression()
model2 = lm2.fit(X_train2,y_train2)


In [45]:
#rerun predictive set
y_pred2 = model2.predict(X_test2)
pd.DataFrame(y_pred2).to_clipboard()

In [46]:
#print regression stats
rmse = sqrt(mean_squared_error(y_test2, y_pred2))
mae2 = mean_absolute_error(y_test2, y_pred2)
evs2 = explained_variance_score(y_test2, y_pred2)

print(lm2.intercept_)
print(mae2)
print(evs2)


64.08978920438395
5.523311392028644
0.9236919464675655


In [47]:
coef2 = np.round(lm2.coef_,2)

In [48]:
df_Coef = pd.DataFrame(abs(coef2), X2.columns, columns=['Coefficient']).sort_values(by='Coefficient', ascending = False)

df_Coef.to_clipboard()
df_Coef

Coefficient
AdultWeekday             19.91
summit_elev              13.34
base_elev                11.31
vertical_drop             4.96
Runs                      1.64
quad                      1.54
averageSnowfall           1.44
triple                    1.41
cluster                   1.37
surface                   1.33
daysOpenLastYear          0.87
fastQuads                 0.76
total_chairs              0.69
NightSkiing_ac            0.53
Snow Making_ac            0.48
fastSixes                 0.42
SkiableTerrain_ac         0.37
double                    0.35
projectedDaysOpen         0.31
yearsOpen                 0.29
LongestRun_mi             0.29
trams                     0.27
TerrainParks              0.23
fastEight                 0.02

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [49]:
#drop state, name, adulstweekend
X = df.drop(['Name','AdultWeekend', 'state', 'Region', 'summit_elev', 'base_elev', 'vertical_drop'], axis=1)
#set y
y= df.AdultWeekend 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train3, X_test3, y_train3, y_test3 = train_test_split(X_scaled,y, test_size=0.25, random_state=1)

In [50]:
lm3 = linear_model.LinearRegression()
model3 = lm3.fit(X_train3,y_train3)


In [51]:
y_pred3 = model3.predict(X_test3)

In [52]:
rmse3 = sqrt(mean_squared_error(y_test3, y_pred3))
mae3 = mean_absolute_error(y_test3, y_pred3)
evs3 = explained_variance_score(y_test3, y_pred3)

print(model3.intercept_)
print(mae3)
print(evs3)

64.07623856428401
5.3686412085661415
0.9300472880181068


In [53]:
coef3 = np.round(lm3.coef_,2)

In [54]:
pd.DataFrame(abs(coef3), X.columns, columns=['Coefficient']).sort_values(by='Coefficient', ascending = False).head(10)

Coefficient
AdultWeekday            20.11
surface                  1.37
quad                     1.36
triple                   1.33
averageSnowfall          1.27
cluster                  1.05
daysOpenLastYear         1.03
Runs                     0.99
fastQuads                0.92
total_chairs             0.65

In [55]:
print('Mean Absolute Error')
print(round(mae,2),round(mae2,2),round(mae3,2))
print('Explained Variance')
print(round(evs,4),round(evs2,4),round(evs3,4))




Mean Absolute Error
5.16 5.52 5.37
Explained Variance
0.9307 0.9237 0.93


## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | .9307 |5.16  |-|
| Model 2. | 0.9237|5.52 |'state'|
| Model 3. | 0.93 |5.37 |'state','summit_elev','base_elev'|

Model Selection:

In [56]:
# Additional Analysis to compare Actual vs. 3 predicted scenarios in DF


compdf = pd.DataFrame({'Actual1':y_test, 'Predicted1': y_pred}, columns = ['Actual1','Predicted1'])
compdf2 =  pd.DataFrame({'Predicted2': y_pred2}, columns = ['Predicted2'])
compdf3 = pd.DataFrame({'Predicted3': y_pred3}, columns = ['Predicted3'])

price_comp = pd.concat([compdf, compdf2, compdf3], axis = 1)
price_comp.head(10)


Actual1  Predicted1  Predicted2  Predicted3
0  64.16681   58.865216   58.011228   59.044570
1  88.00000   89.147198   84.729837   85.986789
2  74.00000   78.979718   82.909505   81.974271
3  78.00000   76.421612   78.810327   78.854083
4  64.16681   70.250469   67.523225   67.500772
5  79.00000   76.024639   81.233366   81.168481
6  60.00000   44.856671   38.038650   37.059827
7  69.00000   46.667462   47.757783   47.559765
8  81.00000   93.522320   88.799709   88.469378
9  88.00000   83.486553   85.353500   85.781994